### PR TITLE
process replay: automatically push refs on fail

### PR DESCRIPTION
**Pull request by sshane**
Idea: if a process replay job fails, and the diff is expected, just copy the printed commit to ref_commit and push! 

Todo:
- [x] ~Add azure-cli to docker~ - only used for token generation, can skip this
- [x] Add azure to python env
- [x] Clean up
- [x] Upload!

Quirks (features?):
- Instead of using the pull request HEAD commit, `actions/checkout` merges the PR into the base branch, so the ref_commit that is used is something you don't know until process_replay finishes, meaning you can't: push a commit that has a diff and then immediately push another commit with the previous commit's hash to update refs

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
> Idea: if a process replay job fails, and the diff is expected, just copy the printed commit to ref_commit and push!

How can you know if the diff is expected? Updating the ref commit should always be a manual, intentional process.

**Comment by sshane** [comment]
Well you check the diff first of course, then manually update the ref_commit from the CI workflow

**Comment by adeebshihadeh** [comment]
Ah, misunderstood. I thought you meant updating the commit automatically too by looking at the diff or something.

File: selfdrive/test/openpilotci.py [hunk 1]
```diff
@@ -4,19 +4,22 @@
...
 
+
 def upload_file(path, name):
   from azure.storage.blob import BlockBlobService  # pylint: disable=import-error
 
   sas_token = None
   if os.path.isfile(TOKEN_PATH):
     sas_token = open(TOKEN_PATH).read().strip()
+  elif "AZURE_TOKEN" in os.environ:
+    sas_token = os.environ["AZURE_TOKEN"]
```
Comment (adeebshihadeh) [inline selfdrive/test/openpilotci.py:22]:
```suggestion
  sas_token = os.environ.get("AZURE_TOKEN", None)
  if os.path.isfile(TOKEN_PATH):
    sas_token = open(TOKEN_PATH).read().strip()
```

**Comment by sshane** [comment]
- `./test_processes.py`: identical to behavior before, except we now save the new logs locally, with or without a diff
- `./test_processes.py --upload-only`: after you run the above command, instead of `./update_refs.py`, run this to just upload the generated log messages and update the `ref_commit.py`
- `./test_processes.py --update-refs`: runs on all segments, saves logs, uploads, and updates the `ref_commit` file after run. Identical to `update_refs.py`, except we upload after each segment.

File: selfdrive/test/process_replay/test_processes.py [hunk 2]
```diff
@@ -106,47 +106,58 @@ def format_diff(results, ref_commit):
...
+  parser.add_argument("--update-refs", action="store_true",
+                      help="Updates reference logs using current commit")
+  parser.add_argument("--upload-only", action="store_true",
+                      help="Skips testing processes and uploads logs from previous test run")
   args = parser.parse_args()
 
-  cars_whitelisted = len(args.whitelist_cars) > 0
-  procs_whitelisted = len(args.whitelist_procs) > 0
+  full_test = 0 == len(args.whitelist_procs) == len(args.whitelist_cars) == len(args.blacklist_procs) == \
+              len(args.blacklist_cars) == len(args.ignore_fields) == len(args.ignore_msgs)
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/test_processes.py:133]:
```suggestion
  full_test = all(len(x) == 0 for x in (args.whitelist_procs, args.whitelist_cars, args.black_list_procs, args.blcklist_cars, args.ignore_fields, args.ignore_msgs))
```

File: selfdrive/test/process_replay/test_processes.py [hunk 3]
```diff
@@ -157,23 +168,40 @@ def format_diff(results, ref_commit):
...
+  with open(os.path.join(PROC_REPLAY_DIR, "diff.txt"), "w") as f:
     f.write(diff2)
   print(diff1)
 
   if failed:
     print("TEST FAILED")
-    print("\n\nTo update the reference logs for this test run:")
-    print("./update_refs.py")
+    if not upload:
+      print("\n\nTo update the reference logs for this test run:")
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/test_processes.py:197]:
```suggestion
      print("\n\nTo push the new reference logs for this commit run:")
```

File: .github/workflows/selfdrive_tests.yaml [hunk 4]
```diff
@@ -316,6 +317,11 @@ jobs:
         ${{ env.RUN }} "scons -j$(nproc) && \
                         FILEREADER_CACHE=1 CI=1 coverage run selfdrive/test/process_replay/test_processes.py && \
                         coverage xml"
+    - name: Upload reference logs
+      if: ${{ failure() && github.event_name == 'pull_request' && github.repository == 'commaai/openpilot' && env.AZURE_TOKEN != '' }}
```
[the remark this answers is not in the document]
Comment (sshane) [reply .github/workflows/selfdrive_tests.yaml:321]: so it won't run on test branches and upload needlessly, like e2e_long for example

File: selfdrive/test/process_replay/test_processes.py [hunk 5]
```diff
@@ -106,47 +106,57 @@ def format_diff(results, ref_commit):
...
+                      help="Msgs to ignore (e.g. carEvents)")
+  parser.add_argument("--update-refs", action="store_true",
+                      help="Updates reference logs using current commit")
+  parser.add_argument("--upload-only", action="store_true",
+                      help="Skips testing processes and uploads logs from previous test run")
   args = parser.parse_args()
 
-  cars_whitelisted = len(args.whitelist_cars) > 0
-  procs_whitelisted = len(args.whitelist_procs) > 0
+  full_test = all(len(x) == 0 for x in (args.whitelist_procs, args.whitelist_cars, args.black_list_procs, args.blcklist_cars, args.ignore_fields, args.ignore_msgs))
```
Comment (sshane) [inline selfdrive/test/process_replay/test_processes.py:132]:
```suggestion
  full_test = all(len(x) == 0 for x in (args.whitelist_procs, args.whitelist_cars, args.blacklist_procs, args.blcklist_cars, args.ignore_fields, args.ignore_msgs))
```

File: selfdrive/test/process_replay/test_processes.py [hunk 6]
```diff
@@ -106,47 +106,57 @@ def format_diff(results, ref_commit):
...
+                      help="Msgs to ignore (e.g. carEvents)")
+  parser.add_argument("--update-refs", action="store_true",
+                      help="Updates reference logs using current commit")
+  parser.add_argument("--upload-only", action="store_true",
+                      help="Skips testing processes and uploads logs from previous test run")
   args = parser.parse_args()
 
-  cars_whitelisted = len(args.whitelist_cars) > 0
-  procs_whitelisted = len(args.whitelist_procs) > 0
+  full_test = all(len(x) == 0 for x in (args.whitelist_procs, args.whitelist_cars, args.blacklist_procs, args.blcklist_cars, args.ignore_fields, args.ignore_msgs))
```
Comment (sshane) [inline selfdrive/test/process_replay/test_processes.py:132]:
```suggestion
  full_test = all(len(x) == 0 for x in (args.whitelist_procs, args.whitelist_cars, args.blacklist_procs, args.blacklist_cars, args.ignore_fields, args.ignore_msgs))
```

**Pull request merged**